src/operator/tensor/init_op.h

```diff
@@ -11,20 +11,89 @@
 #include <mxnet/op_attr_types.h>
 #include <dmlc/parameter.h>
 #include <vector>
+#include <string>
+#include <limits>
 #include "../elemwise_op_common.h"
 
 namespace mxnet {
 namespace op {
 
 struct InitOpParam : public dmlc::Parameter<InitOpParam> {
   TShape shape;
+  std::string ctx;
+  int dtype;
   DMLC_DECLARE_PARAMETER(InitOpParam) {
     DMLC_DECLARE_FIELD(shape)
     .set_default(TShape())
     .describe("The shape of the output");
+    DMLC_DECLARE_FIELD(ctx)
+    .set_default("")
+    .describe("Context of output, in format [cpu|gpu|cpu_pinned](n)."
+              "Only used for imperative calls.");
+    DMLC_DECLARE_FIELD(dtype).set_default(mshadow::kFloat32)
+    .add_enum("float32", mshadow::kFloat32)
+    .add_enum("float64", mshadow::kFloat64)
+    .add_enum("float16", mshadow::kFloat16)
+    .add_enum("uint8", mshadow::kUint8)
+    .add_enum("int32", mshadow::kInt32)
+    .describe("Target data type.");
   }
 };
 
+struct RangeParam : public dmlc::Parameter<RangeParam> {
+  real_t start;
+  real_t stop;
+  real_t step;
+  int repeat;
+  std::string ctx;
+  int dtype;
+  DMLC_DECLARE_PARAMETER(RangeParam) {
+    DMLC_DECLARE_FIELD(start)
+    .set_default(std::numeric_limits<real_t>::infinity())
```

Comment:
Could you try dmlc::optional<int>? It has been merged. Just checkout dmlc-core master